File: voss/harness/session.py
"""Persisted session snapshots.

Sessions live at <cwd>/.voss/sessions/<id>.json. Legacy pre-M2 sessions remain
readable in place at $XDG_STATE_HOME/voss/sessions/ but are never written to.
Each snapshot stores the episodic transcript, cwd, model, total cost, and a
per-turn list of RunRecords.

Storage location
----------------
Sessions live at <cwd>/.voss/sessions/<id>.json. Legacy pre-M2 sessions remain
readable in place at $XDG_STATE_HOME/voss/sessions/ but are never written to.

Redaction guarantee
-------------------
SessionRecord is a fixed-field dataclass. Save serializes via dataclasses.asdict,
which means nothing outside the schema gets written. Provider credentials
(API keys, OAuth access/refresh tokens, Bearer headers, anthropic-beta marker)
are NEVER fields on this record and therefore cannot be saved.

User-provided prompt text is allowed to contain anything — including strings
that look like secrets — because EpisodicMemory.content is part of the
allowlist by design (the user typed it). The guarantee is specifically about
what the harness itself attaches to the record (it attaches nothing
secret-shaped).

This invariant is enforced at build time by tests/harness/test_session_redaction.py.
Adding a new SessionRecord field that could carry creds is a breaking change
and must be paired with an explicit redaction step.

RunRecord follows the same fixed-field allowlist. Adding a RunRecord field
that could carry creds is a breaking change and must be paired with an
explicit redaction step. The invariant is enforced by
tests/harness/test_session_redaction.py over both SessionRecord and RunRecord
field values.

M9-06 added two additive fields — `parent_id` (Optional[str], UUID-shaped)
and `parent_turn_index` (Optional[int], non-negative). Neither can carry
provider credentials. Backward compat: `_hydrate` already filters unknown
keys, so old reader + new file silently drops them, and new reader + old
file falls back to the None defaults.
"""
from __future__ import annotations

import dataclasses
import json
import os
import uuid
import warnings
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from voss_runtime import EpisodicMemory


def _sessions_dir(cwd: Path) -> Path:
    return (cwd / ".voss" / "sessions").resolve()


def _legacy_state_dir() -> Path:
    base = Path(os.environ.get("XDG_STATE_HOME", Path.home() / ".local" / "state"))
    return base / "voss" / "sessions"


# Public alias for callers outside this module (cli.doctor_cmd).
legacy_state_dir = _legacy_state_dir


# T1-01: Exit reason vocabulary for the iteration loop (SPEC ITER-06).
# Single source of truth — RunRecord __post_init__ enforces membership.
# T2-03: extended with "batch-invariant" (PAR-02) — 5th additive value
# surfaced when the partition scheduler raises BatchInvariantError.
EXIT_REASONS: frozenset[str] = frozenset(
    {"done", "max-iter", "budget", "interrupt", "batch-invariant", "timeout"}
)
# O3 OBRD-09 / R-04: "timeout" added for forced-timeout cards (additive).


@dataclass
class BatchRecord:
    """One parallel read-batch within an iteration (T2-01, PAR-06).

    Mutating singletons emit no BatchRecord (SPEC PAR-06 line 67). Nested
    inside IterationRecord.batches so each batch belongs to the iteration
    that produced it (T2-CONTEXT D-08).
    """

    batch_index: int
    step_indices: list[int] = field(default_factory=list)
    parallel_count: int = 0
    wall_clock_ms: int = 0
    ok_count: int = 0
    err_count: int = 0


@dataclass
class IterationRecord:
    """One iteration of the agent loop. Persisted under RunRecord.iterations."""

    index: int
    plan: dict = field(default_factory=dict)
    tool_results: list[dict] = field(default_factory=list)
    cost_usd: float = 0.0
    prompt_tokens: int = 0
    completion_tokens: int = 0
    # T4 CACHE-07 (Pitfall 8 / Open Question 3): additive defaults preserve pre-T4 session JSON round-trip.
    cache_creation_input_tokens: int = 0
    cache_read_input_tokens: int = 0
    started_at: str = ""
    ended_at: str = ""
    exit_reason: Optional[str] = None
    batches: list[BatchRecord] = field(default_factory=list)


@dataclass
class RunRecord:
    id: str
    started_at: str
    ended_at: str
    goal: str = ""
    plan: Optional[dict] = None
    inspected: list[str] = field(default_factory=list)
    changed: list[str] = field(default_factory=list)
    avoided: list[dict] = field(default_factory=list)
    assumptions: list[str] = field(default_factory=list)
    decisions: list[dict] = field(default_factory=list)
    risks: list[str] = field(default_factory=list)
    validation: list[dict] = field(default_factory=list)
    failures: list[dict] = field(default_factory=list)
    diff_summary: str = ""
    follow_ups: list[str] = field(default_factory=list)
    cost_usd: float = 0.0
    # T1-01: additive iteration-loop fields. Defaults preserve pre-T1 JSON
    # round-trip and `voss resume` behavior unchanged from v0.1.
    iterations: list[IterationRecord] = field(default_factory=list)
    iteration_count: int = 0
    exit_reason: Optional[str] = None
    iteration_total_prompt_tokens: int = 0
    iteration_total_completion_tokens: int = 0
    # M15-05: skill audit events (additive, defaults for back-compat)
    skill_events: list[dict] = field(default_factory=list)
    scope_denials: list[dict] = field(default_factory=list)

    def __post_init__(self) -> None:
        if self.exit_reason is not None and self.exit_reason not in EXIT_REASONS:
            raise ValueError(
                f"invalid exit_reason {self.exit_reason!r}; "
                f"must be one of {sorted(EXIT_REASONS)}"
            )


@dataclass
class SessionRecord:
    id: str
    name: str
    cwd: str
    model: str
    started_at: str
    updated_at: str
    total_cost_usd: float = 0.0
    turns: list[dict] = field(default_factory=list)
    runs: list[dict] = field(default_factory=list)
    # M9-06 fork lineage. Additive Optional; pre-M9 sessions hydrate as None.
    parent_id: Optional[str] = None
    parent_turn_index: Optional[int] = None

    @classmethod
    def new(cls, *, cwd: Path, model: str, name: str = "") -> "SessionRecord":
        sid = uuid.uuid4().hex[:12]
        now = datetime.now(timezone.utc).isoformat(timespec="seconds")
        return cls(
            id=sid,
            name=name or f"session-{sid[:8]}",
            cwd=str(cwd.resolve()),
            model=model,
            started_at=now,
            updated_at=now,
        )

    def first_task(self) -> str:
        for t in self.turns:
            if t.get("role") == "user":
                return t.get("content", "")[:60]
        return "(empty)"


_SESSION_FIELDS = {f.name for f in dataclasses.fields(SessionRecord)}


def _hydrate(data: dict) -> SessionRecord:
    kept = {k: v for k, v in data.items() if k in _SESSION_FIELDS}
    kept.setdefault("turns", [])
    kept.setdefault("runs", [])
    return SessionRecord(**kept)


def session_path(session_id: str, cwd: Optional[Path] = None) -> Path:
    if cwd is None:
        warnings.warn(
            "session_path() without cwd is deprecated; pass cwd explicitly",
            DeprecationWarning,
            stacklevel=2,
        )
        cwd = Path.cwd()
    return _sessions_dir(cwd) / f"{session_id}.json"


def save(record: SessionRecord, history: EpisodicMemory) -> Path:
    record.updated_at = datetime.now(timezone.utc).isoformat(timespec="seconds")
    record.turns = history.last(10_000)  # full transcript
    cwd = Path(record.cwd)
    path = _sessions_dir(cwd) / f"{record.id}.json"
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(asdict(record), indent=2))
    path.chmod(0o600)
    return path


def _scan_dir(dir_path: Path, session_id_or_name: str) -> list[tuple[dict, Path]]:
    out: list[tuple[dict, Path]] = []
    if not dir_path.exists():
        return out
    for p in dir_path.glob("*.json"):
        try:
            data = json.loads(p.read_text())
        except (OSError, json.JSONDecodeError):
            continue
        if data.get("id", "").startswith(session_id_or_name) or data.get("name") == session_id_or_name:
            out.append((data, p))
    return out


def load(
    session_id_or_name: str,
    cwd: Optional[Path] = None,
) -> tuple[SessionRecord, EpisodicMemory]:
    """Resolve by id prefix or name, return record + rehydrated memory.

    Scans <cwd>/.voss/sessions first, then falls back to the legacy XDG dir.
    Legacy hits are tagged via `_legacy = True`.
    """
    primary: list[tuple[dict, Path]] = []
    if cwd is not None:
        primary = _scan_dir(_sessions_dir(cwd), session_id_or_name)
    legacy = _scan_dir(_legacy_state_dir(), session_id_or_name)

    if primary:
        matches = primary
        is_legacy = False
    elif legacy:
        matches = legacy
        is_legacy = True
    else:
        raise FileNotFoundError(f"no session: {session_id_or_name}")

    if len(matches) > 1:
        names = ", ".join(m[0]["id"][:8] for m in matches)
        raise ValueError(f"ambiguous session id; candidates: {names}")

    data = matches[0][0]
    record = _hydrate(data)
    if is_legacy:
        record._legacy = True  # type: ignore[attr-defined]
    history = EpisodicMemory(capacity=40)
    for t in record.turns:
        history.add(t.get("content", ""), role=t.get("role", "user"))
    return record, history


def list_sessions(
    cwd: Path,
    *,
    include_legacy: bool = False,
) -> list[SessionRecord]:
    out: list[SessionRecord] = []
    primary_dir = _sessions_dir(cwd)
    if primary_dir.exists():
        for p in primary_dir.glob("*.json"):
            try:
                data = json.loads(p.read_text())
            except (OSError, json.JSONDecodeError):
                continue
            out.append(_hydrate(data))

    if include_legacy:
        legacy_dir = _legacy_state_dir()
        if legacy_dir.exists():
            for p in legacy_dir.glob("*.json"):
                try:
                    data = json.loads(p.read_text())
                except (OSError, json.JSONDecodeError):
                    continue
                rec = _hydrate(data)
                rec._legacy = True  # type: ignore[attr-defined]
                out.append(rec)

    out.sort(key=lambda r: r.updated_at, reverse=True)
    return out


def delete(session_id: str, cwd: Optional[Path] = None) -> bool:
    if cwd is None:
        cwd = Path.cwd()
    primary_dir = _sessions_dir(cwd)
    if primary_dir.exists():
        for p in primary_dir.glob("*.json"):
            try:
                data = json.loads(p.read_text())
            except (OSError, json.JSONDecodeError):
                continue
            if data.get("id", "").startswith(session_id) or data.get("name") == session_id:
                p.unlink()
                return True
    # Refuse to delete legacy files.
    legacy_dir = _legacy_state_dir()
    if legacy_dir.exists():
        for p in legacy_dir.glob("*.json"):
            try:
                data = json.loads(p.read_text())
            except (OSError, json.JSONDecodeError):
                continue
            if data.get("id", "").startswith(session_id) or data.get("name") == session_id:
                warnings.warn(
                    f"refusing to delete legacy session at {p}; legacy dir is read-only",
                    DeprecationWarning,
                    stacklevel=2,
                )
                return False
    return False

File: crates/voss-cli/src/session.rs
//! Persisted session snapshots. Wire-format-compatible with
//! `voss/harness/session.py`.
//!
//! Storage path: `$XDG_STATE_HOME/voss/sessions/<id>.json` (default
//! `~/.local/state/voss/sessions/`).
//!
//! IMPORTANT: provider creds (access tokens, refresh tokens, API keys) are
//! NEVER serialized into a session record.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

fn state_dir() -> PathBuf {
    let base = std::env::var("XDG_STATE_HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|_| {
            dirs::home_dir()
                .unwrap_or_default()
                .join(".local")
                .join("state")
        });
    base.join("voss").join("sessions")
}

pub fn session_path(id: &str) -> PathBuf {
    state_dir().join(format!("{id}.json"))
}

/// Mirrors `voss/harness/session.py::SessionRecord` field-for-field.
/// JSON key order: id, name, cwd, model, started_at, updated_at, total_cost_usd, turns.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct SessionRecord {
    pub id: String,
    pub name: String,
    pub cwd: String,
    pub model: String,
    /// ISO-8601 UTC, seconds precision. Matches Python
    /// `datetime.now(timezone.utc).isoformat(timespec="seconds")`,
    /// e.g. `"2026-05-09T15:30:00+00:00"`.
    pub started_at: String,
    pub updated_at: String,
    #[serde(default)]
    pub total_cost_usd: f64,
    #[serde(default)]
    pub turns: Vec<Turn>,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct Turn {
    pub role: String,
    pub content: String,
    /// Preserves any unknown fields a Python session may carry.
    #[serde(flatten)]
    pub extra: BTreeMap<String, serde_json::Value>,
}

impl SessionRecord {
    pub fn new(cwd: &Path, model: &str, name: Option<&str>) -> Self {
        // 12-hex-char id, matching Python `uuid.uuid4().hex[:12]`.
        let full = uuid::Uuid::new_v4().simple().to_string();
        let id: String = full.chars().take(12).collect();
        let now = iso_now();
        let resolved_cwd = cwd
            .canonicalize()
            .unwrap_or_else(|_| cwd.to_path_buf())
            .to_string_lossy()
            .into_owned();
        let derived_name = format!("session-{}", &id[..id.len().min(8)]);
        Self {
            id,
            name: name
                .map(|n| {
                    if n.is_empty() {
                        derived_name.clone()
                    } else {
                        n.to_string()
                    }
                })
                .unwrap_or(derived_name),
            cwd: resolved_cwd,
            model: model.into(),
            started_at: now.clone(),
            updated_at: now,
            total_cost_usd: 0.0,
            turns: Vec::new(),
        }
    }

    pub fn first_task(&self) -> String {
        for t in &self.turns {
            if t.role == "user" {
                return t.content.chars().take(60).collect();
            }
        }
        "(empty)".into()
    }
}

/// Format current time as Python `isoformat(timespec="seconds")` does:
/// `YYYY-MM-DDTHH:MM:SS+00:00`.
pub fn iso_now() -> String {
    let now: DateTime<Utc> = Utc::now();
    now.format("%Y-%m-%dT%H:%M:%S+00:00").to_string()
}

pub fn save(record: &mut SessionRecord) -> std::io::Result<PathBuf> {
    record.updated_at = iso_now();
    let path = session_path(&record.id);
    if let Some(p) = path.parent() {
        std::fs::create_dir_all(p)?;
    }
    let bytes = serde_json::to_vec_pretty(record)?;
    std::fs::write(&path, bytes)?;
    set_owner_only(&path)?;
    Ok(path)
}

#[cfg(unix)]
fn set_owner_only(path: &Path) -> std::io::Result<()> {
    use std::os::unix::fs::PermissionsExt;
    std::fs::set_permissions(path, std::fs::Permissions::from_mode(0o600))
}

#[cfg(not(unix))]
fn set_owner_only(_path: &Path) -> std::io::Result<()> {
    Ok(())
}

/// Resolve by id-prefix OR exact name. Errors on ambiguity / not-found.
pub fn load(id_or_name: &str) -> std::io::Result<SessionRecord> {
    let dir = state_dir();
    if !dir.exists() {
        return Err(std::io::Error::new(
            std::io::ErrorKind::NotFound,
            format!("no session: {id_or_name}"),
        ));
    }
    let mut matches: Vec<SessionRecord> = Vec::new();
    for entry in std::fs::read_dir(&dir)? {
        let entry = entry?;
        if !entry.path().extension().map_or(false, |e| e == "json") {
            continue;
        }
        let bytes = std::fs::read(entry.path())?;
        let rec: SessionRecord = match serde_json::from_slice(&bytes) {
            Ok(r) => r,
            Err(_) => continue,
        };
        if rec.id.starts_with(id_or_name) || rec.name == id_or_name {
            matches.push(rec);
        }
    }
    match matches.len() {
        0 => Err(std::io::Error::new(
            std::io::ErrorKind::NotFound,
            format!("no session: {id_or_name}"),
        )),
        1 => Ok(matches.into_iter().next().unwrap()),
        _ => {
            let names: Vec<String> = matches
                .iter()
                .map(|r| r.id.chars().take(8).collect())
                .collect();
            Err(std::io::Error::new(
                std::io::ErrorKind::InvalidInput,
                format!("ambiguous session id; candidates: {}", names.join(", ")),
            ))
        }
    }
}

pub fn list_sessions() -> std::io::Result<Vec<SessionRecord>> {
    let dir = state_dir();
    if !dir.exists() {
        return Ok(Vec::new());
    }
    let mut entries: Vec<(std::time::SystemTime, SessionRecord)> = Vec::new();
    for entry in std::fs::read_dir(&dir)? {
        let entry = entry?;
        if !entry.path().extension().map_or(false, |e| e == "json") {
            continue;
        }
        let mtime = entry
            .metadata()
            .and_then(|m| m.modified())
            .unwrap_or(std::time::UNIX_EPOCH);
        let bytes = std::fs::read(entry.path())?;
        if let Ok(rec) = serde_json::from_slice::<SessionRecord>(&bytes) {
            entries.push((mtime, rec));
        }
    }
    entries.sort_by(|a, b| b.0.cmp(&a.0));
    Ok(entries.into_iter().map(|(_, r)| r).collect())
}

pub fn delete(id: &str) -> std::io::Result<bool> {
    let path = session_path(id);
    if path.exists() {
        std::fs::remove_file(&path)?;
        Ok(true)
    } else {
        Ok(false)
    }
}
